flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java

```diff
@@ -0,0 +1,357 @@
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.hours(1L);
+		final JobID jobId = new JobID();
+
+		final AccessExecutionGraph accessExecutionGraph = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		// let's first answer with a JobNotFoundException and then only with the correct result
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(
+			FutureUtils.completedExceptionally(new JobNotFoundException(jobId)),
+			CompletableFuture.completedFuture(accessExecutionGraph));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			try {
+				executionGraphFuture.get();
+
+				fail("The execution graph future should have been completed exceptionally.");
+			} catch (ExecutionException ee) {
+				assertTrue(ee.getCause() instanceof FlinkException);
+			}
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
+		}
+	}
+
+	/**
+	 * Tests that cache entries are cleaned up when their TTL has expired upon
+	 * calling {@link ExecutionGraphCache#cleanup()}.
+	 */
+	@Test
+	public void testCacheEntryCleanup() throws Exception {
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.milliseconds(1L);
+		final JobID jobId1 = new JobID();
+		final JobID jobId2 = new JobID();
+		final AccessExecutionGraph accessExecutionGraph1 = mock(AccessExecutionGraph.class);
+		final AccessExecutionGraph accessExecutionGraph2 = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		when(jobManagerGateway.requestJob(eq(jobId1), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph1));
+		when(jobManagerGateway.requestJob(eq(jobId2), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph2));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+
+			CompletableFuture<AccessExecutionGraph> executionGraph1Future = executionGraphCache.getExecutionGraph(jobId1, jobManagerGateway);
+
+			CompletableFuture<AccessExecutionGraph> executionGraph2Future = executionGraphCache.getExecutionGraph(jobId2, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph1, executionGraph1Future.get());
+
+			assertEquals(accessExecutionGraph2, executionGraph2Future.get());
+
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId1), any(Time.class));
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId2), any(Time.class));
+
+			Thread.sleep(timeToLive.toMilliseconds());
+
+			executionGraphCache.cleanup();
```

Comment:
I'm not sure how this test differentiates from `testExecutionGraphEntryInvalidation`.
Not sure if I'm correct, but the only difference is this `cleanup()` call. So, it could well be that this test can also pass because of the TTL sleep, while `cleanup()` can be no-op.

Maybe we need to expose the internal entries of the cache just for tests, so that we can verify entry cleanup without invoking `getExecutionGraph`

Reply:
You're right. I'll add a `size()` method to solve the problem.